Charts on sixcolors.com flicker when zooming in

shallawa opened the pull request:
#### 8f042c727b90676c0a4f2910e26f4474ff08f6d2
<pre>
Charts on sixcolors.com flicker when zooming in
<a href="https://bugs.webkit.org/show_bug.cgi?id=256620">https://bugs.webkit.org/show_bug.cgi?id=256620</a>
rdar://108930635

Reviewed by Simon Fraser.

Rearrange the logic of RenderBoxModelObject::decodingModeForImageDraw() such that
we call isVisibleInViewport() at the end of this function. But there is only one
exception to this. If the image has the attribute decoding=&quot;async&quot; specified, then
we have to make sure the image will not flicker. And to check that we have to call
isVisibleInViewport().

Fix a subtle one-time-flickering we should not rely on the layer repaint count only
because new layers can be created when pinch zoom the image. In addition to the
repaint count, we can rely on a new flag called hasEverPaintedImages which can be
stored in the NodeRareData. It is initialized to false and it is set to true when
the image is drawn by its RenderObject. An image is allowed to be asynchronously
decoded if layer repaint count is zero and the element&apos;s flag hasEverPaintedImages
is false.

The internal setting setLargeImageAsyncDecodingEnabledForTesting() will be renamed
to setAsyncDecodingEnabledForTesting(). Its use will change to enable async image
decoding for any image regardless of its size.

* LayoutTests/fast/images/async-image-background-change.html:
* LayoutTests/fast/images/async-image-background-image-repeated.html:
* LayoutTests/fast/images/async-image-background-image.html:
* LayoutTests/fast/images/async-image-body-background-image.html:
* LayoutTests/fast/images/async-image-multiple-clients-repaint.html:
* LayoutTests/fast/images/async-image-src-change.html:
* LayoutTests/fast/images/decode-render-static-image.html:
* LayoutTests/fast/images/decoding-attribute-async-small-image.html:
* LayoutTests/fast/images/decoding-attribute-dynamic-async-small-image.html:
* LayoutTests/fast/images/sprite-sheet-image-draw.html:
* LayoutTests/http/tests/images/render-partial-image-load.html:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::hasEverPaintedImages const):
(WebCore::Node::setHasEverPaintedImages):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::hasEverPaintedImages const):
(WebCore::NodeRareData::setHasEverPaintedImages):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setAsyncDecodingEnabledForTesting):
(WebCore::Internals::setLargeImageAsyncDecodingEnabledForTesting): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/265328@main">https://commits.webkit.org/265328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e30b0f2230f6e58b117f11cce5e1fa639d4a94a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12273 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10199 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10641 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10817 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13122 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10787 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11709 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8936 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12676 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9594 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10074 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/9744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13002 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10215 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9368 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9486 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2540 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13633 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -310,36 +310,60 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
+    if (paintInfo.paintBehavior.contains(PaintBehavior::Snapshotting))
         return DecodingMode::Synchronous;
+
+    auto nonFlickeringDecodingMode = [&]() -> DecodingMode {
```
Slightly odd name. I would call this `defaultDecodingMode`

shallawa committed:
Charts on sixcolors.com flicker when zooming in
https://bugs.webkit.org/show_bug.cgi?id=256620
rdar://108930635

Reviewed by Simon Fraser.

Rearrange the logic of RenderBoxModelObject::decodingModeForImageDraw() such that
we call isVisibleInViewport() at the end of this function. But there is only one
exception to this. If the image has the attribute decoding="async" specified, then
we have to make sure the image will not flicker. And to check that we have to call
isVisibleInViewport().

Fix a subtle one-time-flickering we should not rely on the layer repaint count only
because new layers can be created when pinch zoom the image. In addition to the
repaint count, we can rely on a new flag called hasEverPaintedImages which can be
stored in the NodeRareData. It is initialized to false and it is set to true when
the image is drawn by its RenderObject. An image is allowed to be asynchronously
decoded if layer repaint count is zero and the element's flag hasEverPaintedImages
is false.

The internal setting setLargeImageAsyncDecodingEnabledForTesting() will be renamed
to setAsyncDecodingEnabledForTesting(). Its use will change to enable async image
decoding for any image regardless of its size.

* LayoutTests/fast/images/async-image-background-change.html:
* LayoutTests/fast/images/async-image-background-image-repeated.html:
* LayoutTests/fast/images/async-image-background-image.html:
* LayoutTests/fast/images/async-image-body-background-image.html:
* LayoutTests/fast/images/async-image-multiple-clients-repaint.html:
* LayoutTests/fast/images/async-image-src-change.html:
* LayoutTests/fast/images/decode-render-static-image.html:
* LayoutTests/fast/images/decoding-attribute-async-small-image.html:
* LayoutTests/fast/images/decoding-attribute-dynamic-async-small-image.html:
* LayoutTests/fast/images/sprite-sheet-image-draw.html:
* LayoutTests/http/tests/images/render-partial-image-load.html:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::hasEverPaintedImages const):
(WebCore::Node::setHasEverPaintedImages):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::hasEverPaintedImages const):
(WebCore::NodeRareData::setHasEverPaintedImages):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setAsyncDecodingEnabledForTesting):
(WebCore::Internals::setLargeImageAsyncDecodingEnabledForTesting): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/265328@main
